plugins/chat/app/services/chat/create_category_channel.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for creating a new category chat channel.
+  #
+  # @example
+  #  Service::Chat::CreateCategoryChannel.call(
+  #   guardian: guardian,
+  #   name: "SuperChannel",
+  #   description: "This is the best channel",
+  #   slug: "super-channel",
+  #   category_id: category.id,
+  #  )
+  #
+  class CreateCategoryChannel
+    include Service::Base
+
+    # @!method call(guardian:, **params_to_create)
+    #   @param [Guardian] guardian
+    #   @param [Hash] params_to_create
+    #   @option params_to_create [String] name
+    #   @option params_to_create [String] description
+    #   @option params_to_create [String] slug
+    #   @option params_to_create [Boolean] auto_join_users
+    #   @option params_to_create [Integer] category_id
+    #   @return [Service::Base::Context]
+
+    policy :can_create_channel
+    contract
+    step :set_auto_join_users_default
+    policy :category_channel_does_not_exist
+    model :category, :fetch_category
+    model :channel, :create_channel
+    model :membership, :create_membership
+    step :enforce_automatic_channel_memberships
+
+    # @!visibility private
+    class Contract
+      attribute :name, :string
+      attribute :description, :string
+      attribute :slug, :string
+      attribute :category_id, :integer
+      attribute :auto_join_users, :boolean, default: false
+
+      validates :category_id, presence: true
+
+      validates :name, length: { maximum: SiteSetting.max_topic_title_length }
+    end
+
+    private
+
+    def set_auto_join_users_default(contract:, **)
```

Comment:
Not really sure of a nicer way to do this, we have `default_values_from` in `Base` and we use it for `UpdateChannel`, but that is just a way to get the defaults from an existing model. We really need a way to alter the context/params data before it is passed into the contract. I need to do this because sometimes the frontend passes `auto_join_users: ""` which will not default to `false` using the `attribute :auto_join_users, :boolean, default: false` call

Reply:
We can do more or less what you’re doing, but directly in the contract with this:
```ruby
before_validation { self.auto_join_users = auto_join_users.presence || false }
```
Another way I can think of would be to create a custom type (`strict_boolean` for example) that would cast empty values to `false` instead of `nil`.

Otherwise, is it a bug that the frontend provides `""` for `auto_join_users`? If it is, then I guess it should probably be fixed on the FE side and we could add a validation in the contract to ensure `auto_join_users` is either `true` or `false` (with `validates :auto_join_users, inclusion: { in: [true, false] }`).

Reply:
I will add `before_validation`, but you are right the frontend shouldn't send a blank value, will fix